File: manual_processor/workers/progress_do.js
/**
 * Cloudflare Durable Object for WebSocket Progress Tracking
 * This runs in the Workers runtime
 */

export class ProgressDurableObject {
  constructor(state, env) {
    this.state = state;
    this.env = env;
    this.sessions = new Map(); // file_id -> { websockets: Set, state: ProgressState }
  }

  async fetch(request) {
    const url = new URL(request.url);
    const path = url.pathname;
    const method = request.method;

    // Extract file_id from path
    const pathParts = path.split('/').filter(Boolean);
    if (pathParts[0] !== 'progress' || pathParts.length < 2) {
      return new Response(JSON.stringify({ error: 'Invalid path' }), {
        status: 400,
        headers: { 'Content-Type': 'application/json' }
      });
    }

    const fileId = pathParts[1];

    if (method === 'GET') {
      return this.handleGet(fileId);
    } else if (method === 'POST') {
      return this.handlePost(fileId, request);
    } else if (method === 'DELETE') {
      return this.handleDelete(fileId);
    } else if (method === 'GET' && request.headers.get('Upgrade') === 'websocket') {
      return this.handleWebSocket(fileId, request);
    }

    return new Response(JSON.stringify({ error: 'Method not allowed' }), {
      status: 405,
      headers: { 'Content-Type': 'application/json' }
    });
  }

  async handleGet(fileId) {
    // Try to get from storage
    const stored = await this.state.storage.get(`progress:${fileId}`);
    if (!stored) {
      return new Response(JSON.stringify({ error: 'Not found' }), {
        status: 404,
        headers: { 'Content-Type': 'application/json' }
      });
    }

    return new Response(JSON.stringify(stored), {
      headers: { 'Content-Type': 'application/json' }
    });
  }

  async handlePost(fileId, request) {
    let data;
    try {
      data = await request.json();
    } catch {
      return new Response(JSON.stringify({ error: 'Invalid JSON' }), {
        status: 400,
        headers: { 'Content-Type': 'application/json' }
      });
    }

    // Get or create state
    let state = await this.state.storage.get(`progress:${fileId}`);
    if (!state) {
      state = {
        file_id: fileId,
        status: 'pending',
        progress: 0,
        stage: '待機中',
        result: null,
        error: null
      };
    }

    // Update state
    if (data.status) state.status = data.status;
    if (data.progress !== undefined) state.progress = data.progress;
    if (data.stage) state.stage = data.stage;
    if (data.result) state.result = data.result;
    if (data.error) state.error = data.error;

    // Persist
    await this.state.storage.put(`progress:${fileId}`, state);

    // Broadcast to WebSockets
    await this.broadcast(fileId, state);

    return new Response(JSON.stringify({ success: true }), {
      headers: { 'Content-Type': 'application/json' }
    });
  }

  async handleDelete(fileId) {
    await this.state.storage.delete(`progress:${fileId}`);
    const session = this.sessions.get(fileId);
    if (session) {
      // Close all WebSockets
      for (const ws of session.websockets) {
        ws.close();
      }
      this.sessions.delete(fileId);
    }
    return new Response(JSON.stringify({ success: true }), {
      headers: { 'Content-Type': 'application/json' }
    });
  }

  async handleWebSocket(fileId, request) {
    // Upgrade to WebSocket
    const webSocketPair = new WebSocketPair();
    const [client, server] = Object.values(webSocketPair);

    // Accept the WebSocket
    server.accept();

    // Get or create session
    let session = this.sessions.get(fileId);
    if (!session) {
      session = { websockets: new Set(), state: null };
      this.sessions.set(fileId, session);
    }

    session.websockets.add(server);

    // Send current state if available
    const stored = await this.state.storage.get(`progress:${fileId}`);
    if (stored) {
      server.send(JSON.stringify(stored));
    }

    // Handle incoming messages (ping/pong)
    server.addEventListener('message', async (event) => {
      try {
        const data = JSON.parse(event.data);
        if (data.type === 'ping') {
          server.send(JSON.stringify({ type: 'pong' }));
        }
      } catch {
        // Ignore invalid messages
      }
    });

    // Handle close
    server.addEventListener('close', () => {
      session.websockets.delete(server);
      if (session.websockets.size === 0) {
        this.sessions.delete(fileId);
      }
    });

    // Handle errors
    server.addEventListener('error', () => {
      session.websockets.delete(server);
    });

    return new Response(null, {
      status: 101,
      webSocket: client
    });
  }

  async broadcast(fileId, state) {
    const session = this.sessions.get(fileId);
    if (!session) return;

    const message = JSON.stringify(state);
    for (const ws of session.websockets) {
      try {
        ws.send(message);
      } catch {
        // Remove broken connections
        session.websockets.delete(ws);
      }
    }
  }
}

File: manual_processor/functions/_middleware/progress_do.py
"""
Cloudflare Durable Object for WebSocket progress tracking
"""

import json
import logging
from typing import Dict, Set, Optional
from dataclasses import dataclass, asdict

logger = logging.getLogger(__name__)


@dataclass
class ProgressState:
    """Progress state for a file processing job"""
    file_id: str
    status: str = "pending"  # pending, processing, completed, error
    progress: float = 0.0
    stage: str = "待機中"
    result: Optional[Dict] = None
    error: Optional[str] = None
    websockets: Set[str] = None  # Store WebSocket IDs
    
    def __post_init__(self):
        if self.websockets is None:
            self.websockets = set()
    
    def to_dict(self) -> Dict:
        data = asdict(self)
        data["websockets"] = list(self.websockets)
        return data


class ProgressDurableObject:
    """Durable Object for managing WebSocket connections and progress state"""
    
    def __init__(self, state: Any, env: Any):
        self.state = state
        self.env = env
        self.sessions: Dict[str, ProgressState] = {}
    
    async def fetch(self, request: Any) -> Any:
        """Handle HTTP requests to the Durable Object"""
        url = request.url
        method = request.method
        
        # Extract file_id from URL path
        path_parts = url.path.strip('/').split('/')
        if len(path_parts) < 2 or path_parts[0] != "progress":
            return self._json_response({"error": "Invalid path"}, 400)
        
        file_id = path_parts[1]
        
        if method == "GET":
            return await self._handle_get(file_id)
        elif method == "POST":
            return await self._handle_post(file_id, request)
        elif method == "DELETE":
            return await self._handle_delete(file_id)
        else:
            return self._json_response({"error": "Method not allowed"}, 405)
    
    async def _handle_get(self, file_id: str) -> Any:
        """Get progress state"""
        state = self.sessions.get(file_id)
        if not state:
            # Try to load from storage
            stored = await self.state.storage.get(f"progress:{file_id}")
            if stored:
                state = ProgressState(**stored)
                self.sessions[file_id] = state
        
        if not state:
            return self._json_response({"error": "Not found"}, 404)
        
        return self._json_response(state.to_dict())
    
    async def _handle_post(self, file_id: str, request: Any) -> Any:
        """Update progress state"""
        try:
            data = await request.json()
        except Exception:
            return self._json_response({"error": "Invalid JSON"}, 400)
        
        state = self.sessions.get(file_id)
        if not state:
            state = ProgressState(file_id=file_id)
            self.sessions[file_id] = state
        
        # Update state
        if "status" in data:
            state.status = data["status"]
        if "progress" in data:
            state.progress = float(data["progress"])
        if "stage" in data:
            state.stage = data["stage"]
        if "result" in data:
            state.result = data["result"]
        if "error" in data:
            state.error = data["error"]
        
        # Persist to storage
        await self.state.storage.put(f"progress:{file_id}", state.to_dict())
        
        # Broadcast to connected WebSockets
        await self._broadcast(file_id, state.to_dict())
        
        return self._json_response({"success": True})
    
    async def _handle_delete(self, file_id: str) -> Any:
        """Delete progress state"""
        if file_id in self.sessions:
            del self.sessions[file_id]
        await self.state.storage.delete(f"progress:{file_id}")
        return self._json_response({"success": True})
    
    async def _broadcast(self, file_id: str, data: Dict) -> None:
        """Broadcast progress update to all connected WebSockets"""
        state = self.sessions.get(file_id)
        if not state:
            return
        
        # In a real implementation, we'd iterate over WebSocket connections
        # and send the update. For now, we'll just log it.
        logger.info(f"Broadcasting progress for {file_id}: {data['progress']}% - {data['stage']}")
    
    def _json_response(self, data: Dict, status: int = 200) -> Any:
        """Create JSON response"""
        return {
            "status": status,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps(data, ensure_ascii=False)
        }


# WebSocket handler for Pages Functions
class ProgressWebSocket:
    """WebSocket handler for real-time progress updates"""
    
    def __init__(self, env: Any):
        self.env = env
    
    async def handle_websocket(self, request: Any, file_id: str) -> Any:
        """Handle WebSocket connection for progress updates"""
        # Upgrade to WebSocket
        ws_pair = request.env.PROGRESS_DO.fetch(
            f"https://do/websocket/{file_id}",
            method="GET",
            headers={"Upgrade": "websocket"}
        )
        
        # This is a simplified version - actual implementation would
        # use the WebSocket Hibernation API
        return ws_pair


# Export for Pages Functions
export = {
    "ProgressDurableObject": ProgressDurableObject
}